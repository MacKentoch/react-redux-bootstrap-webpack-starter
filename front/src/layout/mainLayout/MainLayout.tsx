import React, { Fragment, useEffect, useCallback, ReactNode } from 'react';
import { withRouter } from 'react-router';
import { RouteComponentProps } from 'react-router';
import NavigationBar from '../../components/navigation';
import BackToTop from '../../components/backToTop/BackToTop';
import navigationModel from '../../config/navigation';
import registerServiceWorker from '../../services/sw/registerServiceWorker';

type Props = {
  children: ReactNode;
} & RouteComponentProps;

function MainLayout({ children }: Props) {
  // #region on mount effect
  useEffect(() => {
    if (typeof window !== undefined) {
      // register service worker (no worry about multiple attempts to register, browser will ignore when already registered)
      registerServiceWorker();
    }
  }, []);
  // #endregion

  // #region callbacks
  /* eslint-disable no-unused-vars*/
  const handleLeftNavItemClick = useCallback(
    (/* event: React.MouseEvent<HTMLButtonElement>, viewName?: string */) => {
      // something to do here?
    },
    [],
  );

  const handleRightNavItemClick = useCallback(
    (/* event: React.MouseEvent<HTMLButtonElement>, viewName?: string */) => {
      // something to do here?
    },
    [],
  );
  // #endregion

  return (
    <Fragment>
      <div id="appContainer">
        <NavigationBar
          brand={navigationModel.brand}
          navModel={navigationModel}
          leftNavItemClick={handleLeftNavItemClick}
          rightNavItemClick={handleRightNavItemClick}
        />
        <div className="container-fluid">{children}</div>
        <BackToTop minScrollY={40} scrollTo={'appContainer'} />
      </div>
    </Fragment>
  );
}

export default withRouter(MainLayout);
