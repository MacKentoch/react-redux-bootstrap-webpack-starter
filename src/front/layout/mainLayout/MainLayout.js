// @flow

import React, { Fragment, useEffect, useCallback } from 'react';
import { withRouter } from 'react-router';
import { type Match, type Location, type RouterHistory } from 'react-router';
import NavigationBar from '../../components/navigation';
import BackToTop from '../../components/backToTop/BackToTop';
import navigationModel from '../../config/navigation';
import registerServiceWorker from '../../services/sw/registerServiceWorker';

// #region flow types
type Props = {
  // from withRouter HOC:
  match: Match,
  location: Location,
  history: RouterHistory,

  children: any,
};
// #endregion

// #region MainLayout HOC
function MainLayout({ history, location, match, children }: Props) {
  // #region on mount effect
  useEffect(() => {
    if (typeof window !== undefined) {
      // register service worker (no worry about multiple attempts to register, browser will ignore when already registered)
      registerServiceWorker();
    }
  }, []);
  // #endregion

  // #region callbacks
  /* eslint-disable no-unused-vars*/
  const handleLeftNavItemClick = useCallback((
    event: SyntheticEvent<>,
    viewName: string,
  ) => {
    // something to do here?
  }, []);

  const handleRightNavItemClick = useCallback((
    event: SyntheticEvent<>,
    viewName: string,
  ) => {
    // something to do here?
  }, []);
  // #endregion

  return (
    <Fragment>
      <div id="appContainer">
        <NavigationBar
          brand={navigationModel.brand}
          navModel={navigationModel}
          handleLeftNavItemClick={handleLeftNavItemClick}
          handleRightNavItemClick={handleRightNavItemClick}
        />
        <div className="container-fluid">{children}</div>
        <BackToTop minScrollY={40} scrollTo={'appContainer'} />
      </div>
    </Fragment>
  );
}

// $FlowIgnore
export default withRouter(MainLayout);
