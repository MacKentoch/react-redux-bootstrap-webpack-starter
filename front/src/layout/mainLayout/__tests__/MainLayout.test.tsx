import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { MemoryRouter } from 'react-router';
import MainLayout from '../index';

// NOTE: we mock Navigation component (we are not testing this one) so we won't need redux provider for MainLayout test anymore
jest.mock('../../../components/navigation');
jest.mock('../../../components/backToTop/BackToTop');
jest.mock('../../../services/sw/registerServiceWorker');

describe('MainLayout component', () => {
  let rootElement: any = null;

  beforeEach(() => {
    rootElement = document.createElement('div');
    document.body.appendChild(rootElement);

    jest.restoreAllMocks();
  });

  afterEach(() => {
    rootElement && document.body.removeChild(rootElement);
    rootElement = null;
  });

  it.only('renders as expected', () => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const NavigationBar = require('../../../components/navigation');
    // Redux connect (shoudl return React component) is "default exported".
    // IMPORTANT: here we mock the return value of connect to mock NavigationBar component:
    NavigationBar.default.mockReturnValueOnce(() => <span>navbar</span>);

    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const BackToTop = require('../../../components/backToTop/BackToTop');
    BackToTop.default.mockImplementationOnce(() => (
      <span>backtotopbutton</span>
    ));

    const { container } = render(
      <ThemeProvider theme={{}}>
        <MemoryRouter>
          <MainLayout>
            <p>children here</p>
          </MainLayout>
        </MemoryRouter>
      </ThemeProvider>,
      rootElement,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
