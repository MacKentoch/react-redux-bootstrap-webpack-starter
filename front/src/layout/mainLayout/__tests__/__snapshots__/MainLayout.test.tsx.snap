// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayout component renders as expected 1`] = `
<div
  id="appContainer"
>
  <div
    class="container-fluid"
  >
    <p>
      children here
    </p>
  </div>
  <span>
    backtotopbutton
  </span>
</div>
`;
